File: HadoopCode/ElkTest/src/main/java/com/itheima/es_hbase/HbaseUtil.java
package com.itheima.es_hbase;

import javafx.scene.control.Tab;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.hbase.HBaseConfiguration;
import org.apache.hadoop.hbase.HColumnDescriptor;
import org.apache.hadoop.hbase.HTableDescriptor;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.client.*;

import java.io.IOException;
import java.util.List;

/**
 * @ProjectName: ElkTest
 * @Package: com.itheima.es_hbase
 * @ClassName: HbaseUtil
 * @Description:
 * @Author: sakura
 * @CreateDate: 2019/8/18 0018 18:30
 * @Version: 1.0
 * 1.获取客户端对象
 * 2.初始化表,判读表是否存在,不存在就需要新建
 * 3.根据rowkey查询数据
 * 4.插入数据
 */

public class HbaseUtil {
    //1、获取客户端的对象
    static Connection connection;

    static {
        Configuration configuration = HBaseConfiguration.create();

        //封装zk
        configuration.set("hbase.zookeeper.quorum","hadoop01:2181,hadoop02:2181,hadoop03:2181");

        try {
            connection = ConnectionFactory.createConnection(configuration);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public  static Table initTable(String tableName,String ...family){
        /**
         * @method: initTable
         * @return: org.apache.hadoop.hbase.client.Table
         * @description:  初始化表，判断是否存在，是就删除，否则就创建Hbas表
         * @Author: sakura
         * @Date: 2019/8/18 0018 18:34
         */
        TableName tblName = TableName.valueOf(tableName);

        Table table =null;

        try {
            //获取admin 对象--------怎么的到的，，，，connect.getAdmin,
            Admin admin = connection.getAdmin();

            //判断表是否存在，
            if(!admin.tableExists(tblName)){

                //构建表的描述器----传入的是表的名字
                HTableDescriptor hTableDescriptor = new HTableDescriptor(tblName);

                //构建列族的描述器---此处用的是循环，，？？？？有很多的列，，多次构建
                for (String colFamily : family) {
                    hTableDescriptor.addFamily(new HColumnDescriptor(colFamily));
                }
                //表不存在，就建表----传入的参数是，表的描述器，根据你的描述创建一个表

                admin.createTable(hTableDescriptor);
                admin.close();
            }
            //获取表-----怎么获取表，只有连接connect能有这个资格
             table = connection.getTable(tblName);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return table;
    }

    public  static String queryByRowkey(String tableName,String family,String columnName,String rowkey){
        /**
         * @method: queryByRowkey
         * @return: java.lang.String
         * @description: 查询数据，给我带来什么的参数信息，我能查到数据呢，
         * ----------------表的名字，列族，列名，rowkey？？？？？
         * @Author: sakura
         * @Date: 2019/8/18 0018 18:48
         */

        //获取表
        Table table = initTable(tableName, family);
        //获取新闻的content
        String content ="";

        try {
            Get get = new Get(rowkey.getBytes());
            Result result = table.get(get);

            //查询指定的列族下，指定列的值
            byte[] value = result.getValue(family.getBytes(), columnName.getBytes());
            //的到的值就是content的内容
            content = new String(value);
        } catch (IOException e) {
            e.printStackTrace();
        }finally {
            try {
                table.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return content;
    }

    public  static void putArticles(List<Article> list){
        /**
         * @method: putArticles
         * @return: void
         * @description: 插入的数据的方法-----批量插入，传入的是一个list,下面有单个插入的方法
         * @Author: sakura
         * @Date: 2019/8/18 0018 18:57
         */
        for (Article article : list) {
            putDataByRowkey("articles", "article", "title", article.getTitle(), article.getId());
            putDataByRowkey("articles", "article", "from", article.getFrom(), article.getId());
            putDataByRowkey("articles", "article", "time", article.getTime(), article.getId());
            putDataByRowkey("articles", "article", "readCount", article.getReadCount(), article.getId());
            putDataByRowkey("articles", "article", "content", article.getContent(), article.getId());
        }





    }


    //新建put方法插入数据的方法
    public static void putDataByRowkey(String tableName, String family, String columnName,
                                       String columnValue, String rowkey) {

        //获取表
        Table table = initTable(tableName, family);
        try {
            Put put = new Put(rowkey.getBytes());
            put.addColumn(family.getBytes(), columnName.getBytes(), columnValue.getBytes());
            table.put(put);
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            try {
                table.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }



}

File: HadoopCode/Flink_Code_211/flink_anli/stream/src/main/scala/cn/itcast/until/HbaseUtil.scala
package cn.itcast.until

import cn.itcast.config.GlobalConfig
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.hbase.{HBaseConfiguration, HColumnDescriptor, HTableDescriptor, TableName}
import org.apache.hadoop.hbase.client._

/**
  * 2019/9/4  纯Hbase 的JavaApi 操作，
  * 1、配置对象（两种方法，配置文件和纯API）
  * 2、链接对象
  * 3、admin
  * 4、CRUD操作
  */
object HbaseUtil {

  //获取配置对象
  private val configuration: Configuration = HBaseConfiguration.create()
  //设置属性
  configuration.set("hbase.zookeeper.quorum", GlobalConfig.hbaseZk) //zk
  configuration.set("hbase.rpc.timeout", GlobalConfig.hbaseRpc) //timeout
  configuration.set("hbase.client.operation.timeout", GlobalConfig.hbaseOperationTimeout) //client的to
  configuration.set("hbase.client.scanner.timeout.period", GlobalConfig.hbaseScanTimeout)

  //获取connection连接------Hbase中的连接工厂（传入配置对象）
  private val connection: Connection = ConnectionFactory.createConnection(configuration)
  //获取admin对象
  private val admin: Admin = connection.getAdmin

  /**
    * 对hbase进行增删改查
    * 1.初始化
    * 2.根据rowkey查询数据
    * 3.根据rowkey插入单列数据
    * 4.根据rowkey插入多列数据
    * 5.根据rowkey删除数据
    */

  /**
    * 1.初始化表----初始化的时候---指定tableName和Family 两个?
    */
  def initTable(tableName: String, family: String): Table = {


    val hbaseTableName: TableName = TableName.valueOf(tableName)
    //先判断表是否存在，存在就直接获取，不存在，需要新建表
    if (!admin.tableExists(hbaseTableName)) {

      //构建表描述器 new HtableXXXX (传入表名)
      val tableDescriptor: HTableDescriptor = new HTableDescriptor(hbaseTableName)
      //构建列族描述器 new HCloumn(传入Family）
      val hColDescriptor: HColumnDescriptor = new HColumnDescriptor(family)
      //表描述器添加列出描述器
      tableDescriptor.addFamily(hColDescriptor)
      //创建表
      admin.createTable(tableDescriptor)
    }

    //获取表，还是链接getTable(tableName)
    val table: Table = connection.getTable(hbaseTableName)
    table
  }

  /**
    * 2.根据rowkey查询数据 table  family column  rowKey
    */
  def queryByRowkey(tableName: String, family: String, columnName: String, rowkey: String): String = {
    //初始化表
    val table: Table = initTable(tableName, family)
    var str = ""
    try {
      val get = new Get(rowkey.getBytes())

      val result: Result = table.get(get)
      //family 的字节和列的字节
      val bytes: Array[Byte] = result.getValue(family.getBytes(), columnName.getBytes())

      if (bytes != null && bytes.size > 0) {
        //非空之后转换成String就是得到的结果
        str = new String(bytes)
      }
    } catch {
      case e: Exception => e.printStackTrace()
    } finally {
      table.close()
    }

    str
  }

  /**
    * 3.根据rowkey插入单列数据 table family colunm value rowkey
    */
  def putDataByRowkey(tableName: String, family: String, columnName: String, columnValue: String, rowkey: String): Unit = {

    //初始化表
    val table: Table = initTable(tableName, family)
    try {
      val put = new Put(rowkey.getBytes())
      put.addColumn(family.getBytes(), columnName.getBytes(), columnValue.getBytes())
      table.put(put)
    } catch {
      case e: Exception => e.printStackTrace()
    } finally {
      table.close()
    }
  }

  /**
    * 4.根据rowkey插入多列数据 table family map rowkey
    */
  def putMapDataByRowkey(tableName: String, family: String, map: Map[String, Any], rowkey: String): Unit = {

    //初始化表
    val table: Table = initTable(tableName, family)
    try {
      val put = new Put(rowkey.getBytes())
      for ((x, y) <- map) {
        put.addColumn(family.getBytes(), x.getBytes(), y.toString.getBytes())
      }
      table.put(put)
    } catch {
      case e: Exception => e.printStackTrace()
    } finally {
      table.close()
    }
  }

  /**
    * 5.根据rowkey删除数据 table  family rowkey
    */
  def delByRowkey(tableName: String, family: String, rowkey: String): Unit = {

    //初始化表
    val table: Table = initTable(tableName, family)
    try {
      val delete = new Delete(rowkey.getBytes())
      delete.addFamily(family.getBytes())
      table.delete(delete)
    } catch {
      case e: Exception => e.printStackTrace()
    } finally {
      table.close()
    }
  }
}


//    * 增删改查测试
/*  def main(args: Array[String]): Unit = {

//    //插入单列数据
    putDataByRowkey("pyg","info","column","11","001")
//    //查询001的数据
    val str: String = queryByRowkey("pyg","info","column","001")
    println(str)

    //插入多列数据
    val map = Map("a"->1,"b"->2)
    putMapDataByRowkey("pyg","info",map,"002")

    //根据rowkey删除数据
    delByRowkey("pyg","info","001")

  }*/
